src/crypto/nss.rs

```diff
@@ -0,0 +1,217 @@
+    fn from(e: rc_crypto::Error) -> Self {
+        BackendError::NSSError(e)
+    }
+}
+
+pub type Result<T> = std::result::Result<T, BackendError>;
+
+fn to_nss_alg(curve: COSEAlgorithm) -> Result<&'static Algorithm> {
+    match curve {
+        // TODO(MS): Are these correct / complete?
+        COSEAlgorithm::ES256
+        | COSEAlgorithm::PS256
+        | COSEAlgorithm::ECDH_ES_HKDF256
+        | COSEAlgorithm::ECDH_SS_HKDF256 => Ok(&ECDH_P256),
+        COSEAlgorithm::ES384 => Ok(&ECDH_P384),
+        x => Err(BackendError::UnsupportedAlgorithm(x)),
+    }
+}
+
+fn to_nss_curve(curve: ECDSACurve) -> Result<Curve> {
+    match curve {
+        // TODO(MS): Are these correct / complete?
+        ECDSACurve::SECP256R1 => Ok(Curve::P256),
+        ECDSACurve::SECP384R1 => Ok(Curve::P256),
+        ECDSACurve::SECP521R1 => Ok(Curve::P384),
+        x => Err(BackendError::UnsupportedCurve(x)),
+    }
+}
+
+/* From CTAP2.1 spec:
+
+initialize()
+
+    This is run by the platform when starting a series of transactions with a specific authenticator.
+encapsulate(peerCoseKey) → (coseKey, sharedSecret) | error
+
+    Generates an encapsulation for the authenticator’s public key and returns the message to transmit and the shared secret.
+encrypt(key, demPlaintext) → ciphertext
+
+    Encrypts a plaintext to produce a ciphertext, which may be longer than the plaintext. The plaintext is restricted to being a multiple of the AES block size (16 bytes) in length.
+decrypt(key, ciphertext) → plaintext | error
+
+    Decrypts a ciphertext and returns the plaintext.
+authenticate(key, message) → signature
+
+    Computes a MAC of the given message.
+*/
+// TODO(MS): Maybe remove ByteBuf and return Vec<u8>'s instead for a cleaner interface
+pub(crate) fn serialize_key(_curve: ECDSACurve, key: &[u8]) -> Result<(ByteBuf, ByteBuf)> {
+    // TODO(MS): I actually have NO idea how to do this with NSS
+    let length = key[1..].len() / 2;
+    let chunks: Vec<_> = key[1..].chunks_exact(length).collect();
+    Ok((
+        ByteBuf::from(chunks[0].to_vec()),
+        ByteBuf::from(chunks[1].to_vec()),
+    ))
+}
+
+pub(crate) fn parse_key(curve: ECDSACurve, x: &[u8], y: &[u8]) -> Result<Vec<u8>> {
+    let nss_name = to_nss_curve(curve)?;
+    // Note:: NSSPublicKey does not provide the from_coordinates-function, so we have to go via EcKey
+    //        and fake a private key.
+    let key =
+        EcKey::from_coordinates(nss_name, &[], x, y).map_err(|e| rc_crypto::Error::from(e))?;
+
+    Ok(key.public_key().to_vec())
+}
+
+/// This is run by the platform when starting a series of transactions with a specific authenticator.
+// pub(crate) fn initialize() {
+//     unimplemented!()
+// }
+
+/// Generates an encapsulation for the authenticator’s public key and returns the message
+/// to transmit and the shared secret.
+pub(crate) fn encapsulate(key: &COSEKey) -> Result<ECDHSecret> {
+    if let COSEKeyType::EC2(ec2key) = &key.key {
```

Comment:
Let's remove the commented openssl implementation here.

Reply:
Whoops! Removed